File: oracle/graph_builder.py
"""Graph Auto-Builder — Proactive Artist Relationship Builder.

Builds the connection graph by running all library artists through the
Lore historian engine on pipeline completion, rather than waiting for
on-demand trace requests.

Two complementary edge types are generated:

lineage edges (via Lore / MusicBrainz):
  member_of, collab, influence, rivalry

dimension_affinity edges (local-only, no API):
  Artists whose track_scores centroids are cosine-similar (>= threshold).
  These edges represent shared emotional DNA — artists who occupy similar
  sonic space in your library, regardless of genre or history.  They are
  cheap to compute, require no external calls, and massively increase graph
  density for constellation and discovery.

Runs incrementally: only processes artists added since the last build run.
Stores progress in the ``meta`` table (key=graph_builder_last_run_ts).

Usage::

    from oracle.graph_builder import GraphBuilder

    gb = GraphBuilder()
    count = gb.build_incremental()
    print(f"Added {count} new connections")

    # Dimension edges only (fast, no rate limiting):
    count = gb.build_dimension_edges()

    # Or run full rebuild (slow on large libraries):
    count = gb.build_full_graph()

Author: Lyra Oracle — Sprint 1
"""

from __future__ import annotations

import json
import logging
import time
from typing import Any, Callable, Dict, List, Optional

import numpy as np

from oracle.db.schema import get_connection

logger = logging.getLogger(__name__)

_META_KEY_LAST_RUN = "graph_builder_last_run_ts"

# Pearson correlation (z-score-normalised cosine) threshold for dimension_affinity edges.
# With raw cosine on absolute CLAP scores, 0.82 would include nearly every pair because
# all scores are positive and cluster in the same direction.  After z-score standardisation
# the metric is equivalent to Pearson correlation and 0.60 gives a meaningful neighbourhood
# of similar sonic profiles without drowning the graph.
# Override with LYRA_GRAPH_DIMENSION_THRESHOLD in .env.
_DIMENSION_SIM_THRESHOLD = float(
    __import__("os").getenv("LYRA_GRAPH_DIMENSION_THRESHOLD", "0.60")
)
_DIMENSIONS = ["energy", "valence", "tension", "density", "warmth", "movement", "space", "rawness", "complexity", "nostalgia"]

logger = logging.getLogger(__name__)

_META_KEY_LAST_RUN = "graph_builder_last_run_ts"


class GraphBuilder:
    """Proactively builds the artist relationship graph.

    Designed to run automatically at the end of ``oracle pipeline``
    and incrementally (only new artists) on subsequent runs.
    """

    def build_full_graph(
        self,
        progress_callback: Optional[Callable[[int, int, str], None]] = None,
        depth: int = 1,
    ) -> int:
        """Build relationships for ALL unique artists in the library.

        This is a full rebuild — slower but thorough. Useful for the
        first run or after a major library import.

        Args:
            progress_callback: Optional callable(current, total, artist_name).
            depth: Genealogy depth passed to Lore.trace_lineage (1–2 recommended).

        Returns:
            Count of new connection edges added to the database.
        """
        artists = self._get_all_unique_artists()
        logger.info("GraphBuilder: full build for %d artists (depth=%d)", len(artists), depth)
        count = self._process_artists(artists, progress_callback=progress_callback, depth=depth)
        count += self.build_dimension_edges()
        return count

    def build_incremental(
        self,
        progress_callback: Optional[Callable[[int, int, str], None]] = None,
        depth: int = 1,
    ) -> int:
        """Build relationships for artists added since the last graph build.

        Much faster than full rebuild — safe to run after every pipeline run.

        Args:
            progress_callback: Optional callable(current, total, artist_name).
            depth: Genealogy depth passed to Lore.trace_lineage.

        Returns:
            Count of new connection edges added.
        """
        last_run_ts = self._get_last_run_ts()
        artists = self._get_new_artists_since(last_run_ts)

        if not artists:
            logger.info("GraphBuilder: no new artists since last run (ts=%.0f)", last_run_ts)
            return 0

        logger.info("GraphBuilder: incremental build for %d new artists (since=%.0f)", len(artists), last_run_ts)
        count = self._process_artists(artists, progress_callback=progress_callback, depth=depth)
        count += self.build_dimension_edges()
        self._set_last_run_ts(time.time())
        return count

    def get_stats(self) -> Dict[str, Any]:
        """Return graph statistics for diagnostics.

        Returns:
            Dict with keys: total_artists, total_connections, last_run_ts, top_connected.
        """
        conn = get_connection()
        try:
            c = conn.cursor()
            c.execute("SELECT COUNT(DISTINCT source) FROM connections")
            total_artists = c.fetchone()[0] or 0

            c.execute("SELECT COUNT(*) FROM connections")
            total_connections = c.fetchone()[0] or 0

            c.execute(
                """
                SELECT source, COUNT(*) as cnt
                FROM connections
                GROUP BY source
                ORDER BY cnt DESC
                LIMIT 5
                """
            )
            top_connected = [{"artist": r[0], "connections": r[1]} for r in c.fetchall()]

            return {
                "total_artists": total_artists,
                "total_connections": total_connections,
                "last_run_ts": self._get_last_run_ts(),
                "top_connected": top_connected,
            }
        finally:
            conn.close()

    def build_dimension_edges(
        self,
        threshold: float = _DIMENSION_SIM_THRESHOLD,
        top_k: int = 8,
    ) -> int:
        """Add ``dimension_affinity`` edges between artists with similar sonic profiles.

        Uses per-artist centroid vectors computed from ``track_scores``.
        No external API calls — purely local DB data.

        Scores are z-score standardised across all artists before comparison
        (equivalent to Pearson correlation) so that the metric captures relative
        profile shape, not absolute score level.  Without standardisation,
        every artist looks similar because all CLAP scores are positive and
        cluster in the same quadrant.

        For each artist the *top_k* most-similar neighbours are kept.  Only
        pairs whose standardised cosine similarity >= *threshold* AND who don't
        already have a ``dimension_affinity`` edge receive a new edge.

        Args:
            threshold: Minimum standardised cosine similarity (default 0.82).
            top_k: Maximum neighbours per artist to consider (default 8).

        Returns:
            Number of new edges inserted.
        """
        conn = get_connection()
        try:
            c = conn.cursor()
            cols = ", ".join(f"AVG(ts.{d})" for d in _DIMENSIONS)
            c.execute(
                f"""
                SELECT t.artist, {cols}
                FROM tracks t
                JOIN track_scores ts ON ts.track_id = t.track_id
                WHERE t.artist IS NOT NULL AND trim(t.artist) != ''
                  AND t.status = 'active'
                GROUP BY t.artist
                HAVING COUNT(ts.track_id) >= 1
                """,
            )
            rows = c.fetchall()
        finally:
            conn.close()

        if len(rows) < 2:
            logger.info("[graph] build_dimension_edges: not enough artist data (%d artists)", len(rows))
            return 0

        artists = [r[0] for r in rows]
        raw_vecs = np.array([[v if v is not None else 0.0 for v in r[1:]] for r in rows], dtype=np.float64)

        # Z-score standardise each dimension across all artists so that similarity
        # reflects shape/pattern of the profile, not absolute level.
        means = raw_vecs.mean(axis=0, keepdims=True)
        stds = raw_vecs.std(axis=0, keepdims=True)
        stds[stds < 1e-9] = 1.0  # guard zero-variance dimensions
        z_vecs = (raw_vecs - means) / stds

        # L2-normalise for cosine similarity
        norms = np.linalg.norm(z_vecs, axis=1, keepdims=True)
        norms[norms < 1e-9] = 1.0
        vecs = z_vecs / norms

        # Pairwise similarity; zero the diagonal so self-matches are excluded
        sim_matrix = (vecs @ vecs.T).astype(np.float32)
        np.fill_diagonal(sim_matrix, 0.0)

        # Load existing dimension_affinity pairs to avoid duplicates
        conn2 = get_connection()
        try:
            c2 = conn2.cursor()
            c2.execute(
                "SELECT source, target FROM connections WHERE type = 'dimension_affinity'"
            )
            existing = {(r[0], r[1]) for r in c2.fetchall()}
        finally:
            conn2.close()

        new_edges: list[tuple] = []
        seen_pairs: set[tuple] = set()
        n = len(artists)

        for i in range(n):
            # top_k neighbours for artist i
            top_indices = np.argpartition(sim_matrix[i], -top_k)[-top_k:]
            for j in top_indices:
                j = int(j)
                if j == i:
                    continue
                sim = float(sim_matrix[i, j])
                if sim < threshold:
                    continue
                a, b = artists[i], artists[j]
                pair = (min(a, b), max(a, b))
                if pair in seen_pairs:
                    continue
                if (a, b) in existing or (b, a) in existing:
                    continue
                seen_pairs.add(pair)

                # strongest shared dimensions for evidence
                avg_z = ((z_vecs[i] + z_vecs[j]) / 2.0)
                top_dims = sorted(
                    zip(_DIMENSIONS, avg_z.tolist()), key=lambda x: abs(x[1]), reverse=True
                )[:3]
                evidence = json.dumps({
                    "similarity": round(sim, 4),
                    "top_dims": [{"dim": d, "zscore": round(float(v), 3)} for d, v in top_dims],
                })
                new_edges.append((a, b, "dimension_affinity", round(sim, 4), evidence))
                new_edges.append((b, a, "dimension_affinity", round(sim, 4), evidence))

        if not new_edges:
            logger.info("[graph] build_dimension_edges: 0 new edges (threshold=%.2f, top_k=%d)", threshold, top_k)
            return 0

        conn3 = get_connection()
        try:
            c3 = conn3.cursor()
            c3.executemany(
                "INSERT INTO connections (source, target, type, weight, evidence) VALUES (?, ?, ?, ?, ?)",
                new_edges,
            )
            conn3.commit()
        finally:
            conn3.close()

        added = len(new_edges) // 2  # bidirectional pairs
        logger.info(
            "[graph] build_dimension_edges: %d new artist pairs (threshold=%.2f, top_k=%d, artists=%d)",
            added, threshold, top_k, n,
        )
        return added

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _get_all_unique_artists(self) -> List[str]:
        """Return sorted list of all unique, non-empty artist names."""
        conn = get_connection()
        try:
            c = conn.cursor()
            c.execute(
                "SELECT DISTINCT artist FROM tracks "
                "WHERE artist IS NOT NULL AND trim(artist) != '' "
                "ORDER BY artist"
            )
            return [row[0].strip() for row in c.fetchall()]
        finally:
            conn.close()

    def _get_new_artists_since(self, ts: float) -> List[str]:
        """Return artists from tracks added after the given Unix timestamp."""
        conn = get_connection()
        try:
            c = conn.cursor()
            # artists in tracks added after ts, excluding those already in connections
            c.execute(
                """
                SELECT DISTINCT t.artist
                FROM tracks t
                WHERE t.artist IS NOT NULL
                  AND trim(t.artist) != ''
                  AND (t.added_at > ? OR t.created_at > ?)
                  AND t.artist NOT IN (SELECT DISTINCT source FROM connections)
                ORDER BY t.artist
                """,
                (ts, ts),
            )
            return [row[0].strip() for row in c.fetchall()]
        finally:
            conn.close()

    def _process_artists(
        self,
        artists: List[str],
        progress_callback: Optional[Callable[[int, int, str], None]],
        depth: int,
    ) -> int:
        """Run Lore.trace_lineage for each artist and count new edges.

        Args:
            artists: List of artist name strings.
            progress_callback: Optional progress reporter.
            depth: Lore trace depth.

        Returns:
            Total new connection rows inserted.
        """
        # Lazy import to avoid circular dependency at module level
        try:
            from oracle.lore import Lore
            lore = Lore()
        except Exception as exc:
            logger.error("GraphBuilder: could not import Lore — %s", exc)
            return 0

        total = len(artists)
        new_edges = 0

        for i, artist_name in enumerate(artists):
            if progress_callback:
                progress_callback(i + 1, total, artist_name)

            before = self._count_connections_for(artist_name)
            try:
                lore.trace_lineage(artist_name, depth=depth)
            except Exception as exc:
                logger.debug("GraphBuilder: trace_lineage failed for '%s': %s", artist_name, exc)
                continue
            after = self._count_connections_for(artist_name)
            new_edges += max(0, after - before)

            # MusicBrainz is rate-limited; respect that
            time.sleep(1.2)

        logger.info("GraphBuilder: done — %d new edges", new_edges)
        return new_edges

    def _count_connections_for(self, artist: str) -> int:
        conn = get_connection()
        try:
            c = conn.cursor()
            c.execute("SELECT COUNT(*) FROM connections WHERE source = ?", (artist,))
            return c.fetchone()[0] or 0
        finally:
            conn.close()

    def _get_last_run_ts(self) -> float:
        """Read last run timestamp from meta table. Returns 0.0 if missing."""
        conn = get_connection()
        try:
            c = conn.cursor()
            c.execute(
                "SELECT value FROM meta WHERE key = ? LIMIT 1",
                (_META_KEY_LAST_RUN,),
            )
            row = c.fetchone()
            return float(row[0]) if row else 0.0
        except Exception:
            return 0.0
        finally:
            conn.close()

    def _set_last_run_ts(self, ts: float) -> None:
        """Persist last run timestamp to meta table."""
        conn = get_connection()
        try:
            c = conn.cursor()
            c.execute(
                "INSERT OR REPLACE INTO meta (key, value) VALUES (?, ?)",
                (_META_KEY_LAST_RUN, str(ts)),
            )
            conn.commit()
        except Exception as exc:
            logger.debug("GraphBuilder: could not write meta — %s", exc)
        finally:
            conn.close()

File: crates/lyra-core/src/graph_builder.rs
//! Artist relationship graph builder — Rust port of oracle/graph_builder.py.
//!
//! Two edge types are supported:
//!
//! **dimension_affinity** (local-only, no API):
//!   Artists whose track_scores centroids are cosine-similar (z-score-normalised).
//!   Implemented in `oracle::build_dimension_affinity` and re-exported here.
//!
//! **similar** (Last.fm community data):
//!   Uses `artist.getSimilar` to surface artists that fans of each library artist
//!   also listen to.  Requires a Last.fm API key stored in provider_configs.
//!
//! **[Lore Lineage?]** — MusicBrainz member_of / influence / lineage edges depend
//! on the Python Lore module and are deferred.
//!
//! Incremental builds track progress via the `settings` KV table
//! (key = `graph_builder_last_run_ts`).

use std::collections::{HashMap, HashSet};

use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{errors::LyraResult, oracle::build_dimension_affinity};

// ── Constants ─────────────────────────────────────────────────────────────────

const LASTFM_BASE: &str = "https://ws.audioscrobbler.com/2.0";
const LASTFM_RATE_MS: u64 = 250; // ~4 req/s, well within 5/s limit
const META_KEY_LAST_RUN: &str = "graph_builder_last_run_ts";

// ── Public types ──────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphEdge {
    pub source: String,
    pub target: String,
    pub edge_type: String,
    pub weight: f64,
    pub evidence: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GraphBuildResult {
    /// Number of new edge pairs (bidirectional) inserted.
    pub new_pairs: usize,
    pub dimension_pairs: usize,
    pub lastfm_pairs: usize,
    pub artists_processed: usize,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GraphStats {
    pub total_artists: usize,
    pub total_connections: usize,
    pub last_run_ts: f64,
    pub top_connected: Vec<GraphNode>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GraphNode {
    pub artist: String,
    pub connections: usize,
}

// ── Credential helpers ────────────────────────────────────────────────────────

pub fn load_lastfm_api_key(conn: &Connection) -> Option<String> {
    conn.query_row(
        "SELECT config_json FROM provider_configs WHERE provider_key = 'lastfm'",
        [],
        |row| row.get::<_, String>(0),
    )
    .ok()
    .and_then(|j| serde_json::from_str::<Value>(&j).ok())
    .and_then(|v| {
        v.get("lastfm_api_key")
            .or_else(|| v.get("LASTFM_API_KEY"))
            .and_then(Value::as_str)
            .map(String::from)
    })
    .filter(|s| !s.is_empty())
}

// ── Meta KV helpers ───────────────────────────────────────────────────────────

fn get_last_run_ts(conn: &Connection) -> f64 {
    conn.query_row(
        "SELECT value_json FROM settings WHERE key = ?1",
        params![META_KEY_LAST_RUN],
        |row| row.get::<_, String>(0),
    )
    .ok()
    .and_then(|s| s.trim_matches('"').parse::<f64>().ok())
    .unwrap_or(0.0)
}

fn set_last_run_ts(conn: &Connection, ts: f64) {
    let _ = conn.execute(
        "INSERT OR REPLACE INTO settings (key, value_json) VALUES (?1, ?2)",
        params![META_KEY_LAST_RUN, ts.to_string()],
    );
}

// ── Last.fm similar-artist API ────────────────────────────────────────────────

#[derive(Debug)]
struct SimilarArtist {
    name: String,
    match_score: f64,
}

/// Call Last.fm `artist.getSimilar` for a single artist.
/// Returns up to `top_k` similar artists with match score ≥ 0.05.
fn fetch_lastfm_similar(api_key: &str, artist: &str, top_k: usize) -> Vec<SimilarArtist> {
    std::thread::sleep(std::time::Duration::from_millis(LASTFM_RATE_MS));

    let data: Value = match ureq::get(LASTFM_BASE)
        .query("method", "artist.getSimilar")
        .query("artist", artist)
        .query("api_key", api_key)
        .query("format", "json")
        .query("limit", &top_k.to_string())
        .call()
    {
        Ok(r) => r.into_json().unwrap_or(Value::Null),
        Err(_) => return vec![],
    };

    data.pointer("/similarartists/artist")
        .and_then(Value::as_array)
        .map(|arr| {
            arr.iter()
                .filter_map(|entry| {
                    let name = entry.get("name").and_then(Value::as_str)?.to_string();
                    // Last.fm returns match as a string like "0.999"
                    let match_score = entry
                        .get("match")
                        .and_then(Value::as_str)
                        .and_then(|s| s.parse::<f64>().ok())
                        .unwrap_or(0.0);
                    if match_score >= 0.05 {
                        Some(SimilarArtist { name, match_score })
                    } else {
                        None
                    }
                })
                .collect()
        })
        .unwrap_or_default()
}

// ── Last.fm similarity edge builder ──────────────────────────────────────────

/// Build `similar` edges from Last.fm community data.
///
/// For each library artist calls `artist.getSimilar` and inserts bidirectional
/// edges.  When `local_targets_only = true` only connects to artists already in
/// the local library (avoids inflating the graph with unowned artists).
///
/// Returns the number of new unique edge pairs inserted.
pub fn build_lastfm_similarity_edges(
    conn: &Connection,
    api_key: &str,
    top_k: usize,
    local_targets_only: bool,
    limit_artists: Option<usize>,
) -> LyraResult<usize> {
    let top_k = top_k.clamp(1, 50);

    // Load library artists
    let library_artists: Vec<String> = {
        let mut stmt = conn.prepare(
            "SELECT DISTINCT ar.name
             FROM artists ar
             JOIN tracks t ON t.artist_id = ar.id
             WHERE trim(COALESCE(ar.name,'')) != ''
               AND (t.quarantined IS NULL OR t.quarantined = 0)
             ORDER BY ar.name",
        )?;
        let all: Vec<String> = stmt
            .query_map([], |row| row.get(0))?
            .filter_map(Result::ok)
            .collect();
        match limit_artists {
            Some(n) => all.into_iter().take(n).collect(),
            None => all,
        }
    };

    let library_set: HashSet<String> = library_artists.iter().cloned().collect();

    // Load existing similar edges to skip duplicates
    let existing: HashSet<(String, String)> = {
        let mut stmt =
            conn.prepare("SELECT source, target FROM connections WHERE type = 'similar'")?;
        let rows_collected: Vec<(String, String)> = stmt
            .query_map([], |row| {
                Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
            })?
            .filter_map(Result::ok)
            .collect();
        rows_collected.into_iter().collect()
    };

    let mut seen_pairs: HashSet<(String, String)> = HashSet::new();
    let mut new_edges: Vec<GraphEdge> = Vec::new();

    for artist_name in &library_artists {
        let similar = fetch_lastfm_similar(api_key, artist_name, top_k);

        for s in similar {
            if s.name.is_empty() || s.name == *artist_name {
                continue;
            }
            if local_targets_only && !library_set.contains(&s.name) {
                continue;
            }
            let lo = if artist_name < &s.name {
                artist_name.clone()
            } else {
                s.name.clone()
            };
            let hi = if artist_name < &s.name {
                s.name.clone()
            } else {
                artist_name.clone()
            };
            let pair = (lo.clone(), hi.clone());
            if seen_pairs.contains(&pair) {
                continue;
            }
            if existing.contains(&(artist_name.clone(), s.name.clone()))
                || existing.contains(&(s.name.clone(), artist_name.clone()))
            {
                continue;
            }
            seen_pairs.insert(pair);

            let weight = (s.match_score * 0.9).clamp(0.0, 1.0);
            let evidence = format!(
                r#"{{"match":{:.4},"source":"lastfm_similar"}}"#,
                s.match_score
            );
            new_edges.push(GraphEdge {
                source: artist_name.clone(),
                target: s.name.clone(),
                edge_type: "similar".to_string(),
                weight,
                evidence: evidence.clone(),
            });
            new_edges.push(GraphEdge {
                source: s.name.clone(),
                target: artist_name.clone(),
                edge_type: "similar".to_string(),
                weight,
                evidence,
            });
        }
    }

    if new_edges.is_empty() {
        return Ok(0);
    }

    let now = chrono::Utc::now().to_rfc3339();
    let mut inserted = 0usize;
    for e in &new_edges {
        let ok = conn.execute(
            "INSERT OR IGNORE INTO connections
             (source, target, type, weight, evidence, updated_at)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
            params![e.source, e.target, e.edge_type, e.weight, e.evidence, now],
        );
        if ok.is_ok() {
            inserted += 1;
        }
    }

    Ok(inserted / 2) // bidirectional pairs
}

// ── Full / incremental build ──────────────────────────────────────────────────

/// Build dimension-affinity + Last.fm similar edges for artists added since the
/// last build run.  Tracks progress in the `settings` table.
///
/// Returns counts of new edge pairs per type.
pub fn build_incremental(
    conn: &Connection,
    top_k_lastfm: usize,
    local_targets_only: bool,
) -> LyraResult<GraphBuildResult> {
    let dim_pairs = build_dimension_affinity(conn);
    let lfm_pairs = match load_lastfm_api_key(conn) {
        Some(key) => {
            build_lastfm_similarity_edges(conn, &key, top_k_lastfm, local_targets_only, None)?
        }
        None => 0,
    };

    set_last_run_ts(conn, chrono::Utc::now().timestamp() as f64);

    // Count artists that were processed (those present in library)
    let artists_processed: i64 = conn
        .query_row(
            "SELECT COUNT(DISTINCT ar.id) FROM artists ar
             JOIN tracks t ON t.artist_id = ar.id
             WHERE (t.quarantined IS NULL OR t.quarantined = 0)",
            [],
            |r| r.get(0),
        )
        .unwrap_or(0);

    Ok(GraphBuildResult {
        new_pairs: dim_pairs + lfm_pairs,
        dimension_pairs: dim_pairs,
        lastfm_pairs: lfm_pairs,
        artists_processed: artists_processed as usize,
    })
}

/// Full rebuild — same as incremental but ignores last-run timestamp.
pub fn build_full(
    conn: &Connection,
    top_k_lastfm: usize,
    local_targets_only: bool,
) -> LyraResult<GraphBuildResult> {
    // Reset timestamp so incremental logic processes everything
    set_last_run_ts(conn, 0.0);
    build_incremental(conn, top_k_lastfm, local_targets_only)
}

// ── Stats ─────────────────────────────────────────────────────────────────────

pub fn graph_stats(conn: &Connection) -> LyraResult<GraphStats> {
    let total_artists: i64 = conn
        .query_row("SELECT COUNT(DISTINCT source) FROM connections", [], |r| {
            r.get(0)
        })
        .unwrap_or(0);

    let total_connections: i64 = conn
        .query_row("SELECT COUNT(*) FROM connections", [], |r| r.get(0))
        .unwrap_or(0);

    let top_connected: Vec<GraphNode> = {
        let mut stmt = conn.prepare(
            "SELECT source, COUNT(*) AS cnt FROM connections
             GROUP BY source ORDER BY cnt DESC LIMIT 5",
        )?;
        let rows_collected: Vec<GraphNode> = stmt
            .query_map([], |row| {
                Ok(GraphNode {
                    artist: row.get(0)?,
                    connections: row.get::<_, i64>(1)? as usize,
                })
            })?
            .filter_map(Result::ok)
            .collect();
        rows_collected
    };

    Ok(GraphStats {
        total_artists: total_artists as usize,
        total_connections: total_connections as usize,
        last_run_ts: get_last_run_ts(conn),
        top_connected,
    })
}

// ── Neighbourhood query ───────────────────────────────────────────────────────

/// Return all direct neighbours of `artist` in the connections table,
/// optionally filtered by edge type.
pub fn get_neighbours(
    conn: &Connection,
    artist: &str,
    edge_type: Option<&str>,
    limit: usize,
) -> LyraResult<Vec<GraphEdge>> {
    let limit = limit.clamp(1, 500);

    let (sql, params_box): (String, Vec<Box<dyn rusqlite::types::ToSql>>) = match edge_type {
        Some(et) => (
            format!(
                "SELECT source, target, type, weight, COALESCE(evidence,'')
                 FROM connections WHERE source = ?1 AND type = ?2
                 ORDER BY weight DESC LIMIT {limit}"
            ),
            vec![Box::new(artist.to_string()), Box::new(et.to_string())],
        ),
        None => (
            format!(
                "SELECT source, target, type, weight, COALESCE(evidence,'')
                 FROM connections WHERE source = ?1
                 ORDER BY weight DESC LIMIT {limit}"
            ),
            vec![Box::new(artist.to_string())],
        ),
    };

    let refs: Vec<&dyn rusqlite::types::ToSql> = params_box.iter().map(|b| b.as_ref()).collect();
    let mut stmt = conn.prepare(&sql)?;
    let edges: Vec<GraphEdge> = stmt
        .query_map(refs.as_slice(), |row| {
            Ok(GraphEdge {
                source: row.get(0)?,
                target: row.get(1)?,
                edge_type: row.get(2)?,
                weight: row.get(3)?,
                evidence: row.get(4)?,
            })
        })?
        .filter_map(Result::ok)
        .collect();

    Ok(edges)
}

// ── Edge type frequency map ───────────────────────────────────────────────────

/// Return a map of edge_type → count across all connections.
pub fn edge_type_counts(conn: &Connection) -> LyraResult<HashMap<String, usize>> {
    let mut stmt = conn.prepare("SELECT type, COUNT(*) FROM connections GROUP BY type")?;
    let map: HashMap<String, usize> = stmt
        .query_map([], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)? as usize))
        })?
        .filter_map(Result::ok)
        .collect();
    Ok(map)
}
